[metricbeat] [gcp] refactor metric filter builder

gpop63 opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What does this PR do?

Extracts the logic to build metrics filters into a function to avoid redundant code.

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
-->

## Why is it important?

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- 

## Use cases

<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

## Screenshots

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

mrg-elastic committed:
refactor getFilterForMetric

mrg-elastic committed:
Merge remote-tracking branch 'upstream/main' into fix_gcp-filter

elasticmachine commented:
## :broken_heart: Tests Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-33560/detail/PR-33560/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-33560/detail/PR-33560/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-33560/detail/PR-33560/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-33560/detail/PR-33560/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_33560.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-11-02T14:47:36.693Z&rangeTo=2022-11-02T15:07:36.693Z&transactionName=BUILD+Beats%2Fbeats%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=fb5afae00a1a569f2af0c6227603c578&transactionId=b2b4d20ce81e8438)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-11-02T14:57:36.693+0000


* Duration: 32 min 19 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 8  |
| Passed       | 1284  |
| Skipped      | 24 |
| Total        | 1316   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->

  ### Test errors [![8](https://img.shields.io/badge/8%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-33560/detail/PR-33560/1//tests)
<details><summary>Expand to view the tests failures</summary><p>
    
    
##### `Build&Test / x-pack/metricbeat-goIntegTest / TestGetFilterForMetric/storage_service_with_region_in_config – github.com/elastic/beats/v7/x-pack/metricbeat/module/gcp/metrics`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestGetFilterForMetric/storage_service_with_region_in_config
    metrics_requester_test.go:186: 
        	Error Trace:	/go/src/github.com/elastic/beats/x-pack/metricbeat/module/gcp/metrics/metrics_requester_test.go:186
        	Error:      	Not equal: 
        	            	expected: "metric.type=\"storage.googleapis.com/firewall/dropped_bytes_count\" AND resource.labels.location = \"us-central1\""
        	            	actual  : "metric.type=\"storage.googleapis.com/firewall/dropped_bytes_count\" AND resource.label.location = starts_with(\"us-central1\")"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-metric.type="storage.googleapis.com/firewall/dropped_bytes_count" AND resource.labels.location = "us-central1"
        	            	+metric.type="storage.googleapis.com/firewall/dropped_bytes_count" AND resource.label.location = starts_with("us-central1")
        	Test:       	TestGetFilterForMetric/storage_service_with_region_in_config
    --- FAIL: TestGetFilterForMetric/storage_service_with_region_in_config (0.00s)
 
 ```
</p></details>
</ul>
    
##### `Build&Test / x-pack/metricbeat-goIntegTest / TestGetFilterForMetric – github.com/elastic/beats/v7/x-pack/metricbeat/module/gcp/metrics`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestGetFilterForMetric
--- FAIL: TestGetFilterForMetric (0.00s)
 
 ```
</p></details>
</ul>
    
##### `Build&Test / x-pack/metricbeat-unitTest / TestGetFilterForMetric/storage_service_with_region_in_config – github.com/elastic/beats/v7/x-pack/metricbeat/module/gcp/metrics`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestGetFilterForMetric/storage_service_with_region_in_config
    metrics_requester_test.go:186: 
        	Error Trace:	/var/lib/jenkins/workspace/PR-33560-1-a8589f12-8325-4e2b-8995-c5273a184f18/src/github.com/elastic/beats/x-pack/metricbeat/module/gcp/metrics/metrics_requester_test.go:186
        	Error:      	Not equal: 
        	            	expected: "metric.type=\"storage.googleapis.com/firewall/dropped_bytes_count\" AND resource.labels.location = \"us-central1\""
        	            	actual  : "metric.type=\"storage.googleapis.com/firewall/dropped_bytes_count\" AND resource.label.location = starts_with(\"us-central1\")"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-metric.type="storage.googleapis.com/firewall/dropped_bytes_count" AND resource.labels.location = "us-central1"
        	            	+metric.type="storage.googleapis.com/firewall/dropped_bytes_count" AND resource.label.location = starts_with("us-central1")
        	Test:       	TestGetFilterForMetric/storage_service_with_region_in_config
    --- FAIL: TestGetFilterForMetric/storage_service_with_region_in_config (0.00s)
 
 ```
</p></details>
</ul>
    
##### `Build&Test / x-pack/metricbeat-unitTest / TestGetFilterForMetric – github.com/elastic/beats/v7/x-pack/metricbeat/module/gcp/metrics`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestGetFilterForMetric
--- FAIL: TestGetFilterForMetric (0.00s)
 
 ```
</p></details>
</ul>
    
##### `Build&Test / x-pack/metricbeat-windows-2022-windows-2022 / TestGetFilterForMetric/storage_service_with_region_in_config – github.com/elastic/beats/v7/x-pack/metricbeat/module/gcp/metrics`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestGetFilterForMetric/storage_service_with_region_in_config
    metrics_requester_test.go:186: 
        	Error Trace:	C:\Users\jenkins\workspace\PR-33560-1-ecbfc8a0-94b5-40c8-91e3-efc3b0ef5d51\src\github.com\elastic\beats\x-pack\metricbeat\module\gcp\metrics\metrics_requester_test.go:186
        	Error:      	Not equal: 
        	            	expected: "metric.type=\"storage.googleapis.com/firewall/dropped_bytes_count\" AND resource.labels.location = \"us-central1\""
        	            	actual  : "metric.type=\"storage.googleapis.com/firewall/dropped_bytes_count\" AND resource.label.location = starts_with(\"us-central1\")"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-metric.type="storage.googleapis.com/firewall/dropped_bytes_count" AND resource.labels.location = "us-central1"
        	            	+metric.type="storage.googleapis.com/firewall/dropped_bytes_count" AND resource.label.location = starts_with("us-central1")
        	Test:       	TestGetFilterForMetric/storage_service_with_region_in_config
    --- FAIL: TestGetFilterForMetric/storage_service_with_region_in_config (0.00s)
 
 ```
</p></details>
</ul>
    
##### `Build&Test / x-pack/metricbeat-windows-2022-windows-2022 / TestGetFilterForMetric – github.com/elastic/beats/v7/x-pack/metricbeat/module/gcp/metrics`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestGetFilterForMetric
--- FAIL: TestGetFilterForMetric (0.00s)
 
 ```
</p></details>
</ul>
    
##### `Build&Test / x-pack/metricbeat-windows-2016-windows-2016 / TestGetFilterForMetric/storage_service_with_region_in_config – github.com/elastic/beats/v7/x-pack/metricbeat/module/gcp/metrics`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestGetFilterForMetric/storage_service_with_region_in_config
    metrics_requester_test.go:186: 
        	Error Trace:	C:\Users\jenkins\workspace\PR-33560-1-635eda41-df9c-4a01-af13-29aa3f90ef27\src\github.com\elastic\beats\x-pack\metricbeat\module\gcp\metrics\metrics_requester_test.go:186
        	Error:      	Not equal: 
        	            	expected: "metric.type=\"storage.googleapis.com/firewall/dropped_bytes_count\" AND resource.labels.location = \"us-central1\""
        	            	actual  : "metric.type=\"storage.googleapis.com/firewall/dropped_bytes_count\" AND resource.label.location = starts_with(\"us-central1\")"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-metric.type="storage.googleapis.com/firewall/dropped_bytes_count" AND resource.labels.location = "us-central1"
        	            	+metric.type="storage.googleapis.com/firewall/dropped_bytes_count" AND resource.label.location = starts_with("us-central1")
        	Test:       	TestGetFilterForMetric/storage_service_with_region_in_config
    --- FAIL: TestGetFilterForMetric/storage_service_with_region_in_config (0.00s)
 
 ```
</p></details>
</ul>
    
##### `Build&Test / x-pack/metricbeat-windows-2016-windows-2016 / TestGetFilterForMetric – github.com/elastic/beats/v7/x-pack/metricbeat/module/gcp/metrics`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestGetFilterForMetric
--- FAIL: TestGetFilterForMetric (0.00s)
 
 ```
</p></details>
</ul>
    
</p></details>


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![13](https://img.shields.io/badge/13%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-33560/detail/PR-33560/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


> Show only the first 10 steps failures

  
  
  
  
##### `x-pack/metricbeat-goIntegTest - mage goIntegTest`
<ul>
<li>Took 2 min 4 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-33560/runs/1/steps/2129/log/?start=0">here</a></li>
<li>Description: <code>mage goIntegTest</code></l1>
</ul>
  
  
  
  
##### `x-pack/metricbeat-goIntegTest - mage goIntegTest`
<ul>
<li>Took 0 min 23 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-33560/runs/1/steps/2207/log/?start=0">here</a></li>
<li>Description: <code>mage goIntegTest</code></l1>
</ul>
  
  
  
  
##### `x-pack/metricbeat-goIntegTest - mage goIntegTest`
<ul>
<li>Took 0 min 23 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-33560/runs/1/steps/2211/log/?start=0">here</a></li>
<li>Description: <code>mage goIntegTest</code></l1>
</ul>
  
  
  
  
##### `x-pack/metricbeat-windows-2022-windows-2022 - mage build unitTest`
<ul>
<li>Took 6 min 19 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-33560/runs/1/steps/2290/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `x-pack/metricbeat-windows-2022-windows-2022 - mage build unitTest`
<ul>
<li>Took 1 min 13 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-33560/runs/1/steps/2795/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `x-pack/metricbeat-windows-2022-windows-2022 - mage build unitTest`
<ul>
<li>Took 1 min 58 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-33560/runs/1/steps/2799/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `x-pack/metricbeat-windows-2016-windows-2016 - mage build unitTest`
<ul>
<li>Took 6 min 29 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-33560/runs/1/steps/2664/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `x-pack/metricbeat-windows-2016-windows-2016 - mage build unitTest`
<ul>
<li>Took 1 min 59 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-33560/runs/1/steps/2916/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `x-pack/metricbeat-windows-2016-windows-2016 - mage build unitTest`
<ul>
<li>Took 1 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-33560/runs/1/steps/2941/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-33560/runs/1/steps/2643/log/?start=0">here</a></li>
<li>Description: <code>Error 'hudson.AbortException: script returned exit code 1'</code></l1>
</ul>
  
  </p>
  </details>



## :bug: Flaky test report
:grey_exclamation: There are test failures but not known flaky tests.



<details><summary>Expand to view the summary</summary>
<p>

### Genuine test errors [![8](https://img.shields.io/badge/8%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-33560/detail/PR-33560/1//tests)
:broken_heart: There are test failures but not known flaky tests, most likely a genuine test failure.

* **Name**: `Build&Test / x-pack/metricbeat-goIntegTest / TestGetFilterForMetric/storage_service_with_region_in_config – github.com/elastic/beats/v7/x-pack/metricbeat/module/gcp/metrics`
* **Name**: `Build&Test / x-pack/metricbeat-goIntegTest / TestGetFilterForMetric – github.com/elastic/beats/v7/x-pack/metricbeat/module/gcp/metrics`
* **Name**: `Build&Test / x-pack/metricbeat-unitTest / TestGetFilterForMetric/storage_service_with_region_in_config – github.com/elastic/beats/v7/x-pack/metricbeat/module/gcp/metrics`
* **Name**: `Build&Test / x-pack/metricbeat-unitTest / TestGetFilterForMetric – github.com/elastic/beats/v7/x-pack/metricbeat/module/gcp/metrics`
* **Name**: `Build&Test / x-pack/metricbeat-windows-2022-windows-2022 / TestGetFilterForMetric/storage_service_with_region_in_config – github.com/elastic/beats/v7/x-pack/metricbeat/module/gcp/metrics`
* **Name**: `Build&Test / x-pack/metricbeat-windows-2022-windows-2022 / TestGetFilterForMetric – github.com/elastic/beats/v7/x-pack/metricbeat/module/gcp/metrics`
* **Name**: `Build&Test / x-pack/metricbeat-windows-2016-windows-2016 / TestGetFilterForMetric/storage_service_with_region_in_config – github.com/elastic/beats/v7/x-pack/metricbeat/module/gcp/metrics`
* **Name**: `Build&Test / x-pack/metricbeat-windows-2016-windows-2016 / TestGetFilterForMetric – github.com/elastic/beats/v7/x-pack/metricbeat/module/gcp/metrics`

</p>
</details>



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

mrg-elastic committed:
fix metrics requester test

endorama reviewed (approved):
~I'm approving this but I'm going to request a second review from the team.~

Just noticed https://github.com/elastic/beats/pull/33728 is not merged yet, once merged will require an update to this PR. Overall the logic looks good to me! (and tests are green)

endorama commented:
https://github.com/elastic/beats/pull/33728 has been merged. @gpop63 may you rebase this?

mrg-elastic committed:
Merge remote-tracking branch 'upstream/main' into fix_gcp-filter

kaiyan-sheng commented on x-pack/metricbeat/module/gcp/metrics/metrics_requester.go:
```diff
@@ -139,111 +164,30 @@ func (r *metricsRequester) getFilterForMetric(serviceName, m string) string {
+	switch serviceName {
+	// Global GCP services that are not regional or zonal.
 	case gcp.ServicePubsub, gcp.ServiceLoadBalancing, gcp.ServiceCloudFunctions, gcp.ServiceFirestore:
 		return f
```
nit: with directly `return f`, we missed the `r.logger.Debugf("ListTimeSeries API filter = %s", f)` debug logging here.

endorama replied:
I slightly refactored the code so that it always log that line in https://github.com/elastic/beats/pull/33560/commits/8549f0bff3964e012a7b3f56b2ea884ede2bdd71

What do you think?

endorama committed:
ensure we are logging applied filter

endorama committed:
fix location detection logic

endorama committed:
fix location detection logic II

Condition required ORing the values as I inverted the condition (it
now checks if config is available, not missing).

endorama committed:
Merge remote-tracking branch 'origin/main' into fix_gcp-filter

endorama committed:
Merge remote-tracking branch 'origin/main' into fix_gcp-filter

aspacca commented on x-pack/metricbeat/module/gcp/metrics/metrics_requester_test.go:
```diff
@@ -85,7 +85,7 @@ func TestGetFilterForMetric(t *testing.T) {
 			"storage.googleapis.com/firewall/dropped_bytes_count",
 			metricsRequester{config: config{Region: "us-central1"}, logger: logger},
-			"metric.type=\"storage.googleapis.com/firewall/dropped_bytes_count\" AND resource.labels.location = \"us-central1\"",
+			"metric.type=\"storage.googleapis.com/firewall/dropped_bytes_count\" AND resource.label.location = starts_with(\"us-central1\")",
```
can we increase coverage in order to include all the scenarios that the refactored method can produce?

endorama committed:
extend tests to cover all possible cases

Refactors tests to improve overall coverage.
I considered impractical and error prone to just add test cases
to the `getFilterForMetric` tests.

To increase coverage and add all possible cases, I refactored all
tests to combine into testing all possible cases.

There have been some minor refactoring to the code too, in order to
make it easier to test but the overall logic did not change.

Multiple tests have been created, addressing the different pieces
composing the filter. The tests now check if a location aware
filter is required, all services default and custom filter labels
applied and all different configuration scenarios (only region,
only zone, only regions or a mix of them).

endorama commented on x-pack/metricbeat/module/gcp/metrics/metrics_requester_test.go:
```diff
@@ -247,6 +75,83 @@ func TestGetTimeIntervalAligner(t *testing.T) {
+	}{
+		{
+			"compute service with empty config",
+			gcp.ServiceCompute,
```
Here `gcp.ServiceCompute` is used as an example for the overall behaviour. All specific behaviours are tested in isolation in other functions:
- `TestBuildRegionsFilter` verifies that the region based filter creation works as intended
- `TestIsAGlobalService` verifies each service (and default) corresponding "global" status flag value
- `TestGetServiceLabelFor` verifies each service (and default) corresponding label value
- `TestTrimWildcard` verifies the wildcard remover produces correct results
- `TestBuildLocationFilter` verifies the location based filter creation works as intended

Keeping tests `TestIsAGlobalService` and `TestGetServiceLabelFor` up to date with the available services and labels in the `gcp` package will ensure all combinations to be covered.
